### ISSUE TITLE
Satellite API Enhancements

### ISSUE THREAD

**Issue by hrueger** [bitfocus/companion#1822]
**Describe the feature**
First of all, thanks for the great new text-based Satellite API. I've just tried it and it works really well.

I'd like to see the following two improvements:
- Send the key type with the `KEY-STATE` event. This way, a client application could know if the key is a `Page Up / Down` or `Page Number` key. Because when it is not receiving / able to draw bitmaps, the key is just black with a blank text.
- Add the current page to the API (get & set)

**Usecases**
- Allow client applications special rendering of the `Page Up / Down` or `Page Number` keys
- Allow client applications to display the current page number as well as to change it. For example, I have a rotary encoder and would like to use that to switch the page. 

I'm happy to create a PR if you just tell me that this is something you would merge :-)

**Comment by Julusian** [comment]
Sending the key type in `KEY-STATE` is a good idea.

I'm not keen on exposing the page number.
I have intentionally kept it out of the api for now, because I think it will add complication and if not thought out well will break some clients if we change how pages are handled inside companion.
Is getting and setting the page number enough? Or will you also want a list of page names?

One big reason is that for 2.3, I intend to look at being able to add, remove and reorder pages. As part of this, I will consider making it so pages aren't identified by a number (because those can change) and are instead identified by a string. Then should satellite clients also use these strings (which will be a breaking api change) or should companion fake page numbers for them?

I'm not saying no, but I think it needs some thought to minimise the chance of having to have a breaking change in a later version of companion

**Comment by Julusian** [comment]
I've added `TYPE` to `KEY-STATE` https://github.com/bitfocus/companion/wiki/Satellite-API#state-change-for-key

**Comment by hrueger** [comment]
Wow, that was fast. Thanks a lot!

I get your point regarding the page concept and I really like the idea of identifying then by their name.

Actually, I don't need the API for the right now and I'm fine waiting a few months. So it would be OK for me to leave this issue open until we got the new page concept and then add that to the api 👍

**Comment by hrueger** [comment]
@Julusian The `type` field works great, however, I believe that something in the API is now broken (maybe due to some refactoring)?

Everything on page 1 works fine, but once I press a `Page Up` or `Page Down` button on my satellite device, I no longer receive any `KEY-STATE` events....

**Comment by Julusian** [comment]
Good spot, was a small typo that is now fixed

**Comment by hrueger** [comment]
Thanks, it's working again.
However, I found another improvement ;-) In the wiki it says that the bitmaps are 72x72 pixels only right now. 
Do you mind if I add a PR which allows a satellite device to specify its button size in the "ADD_DEVICE" message? Should this be completely a number (thus allowing uncommon sizes like 49x49) or rather an enum (the default stream deck button sizes)?

**Comment by hrueger** [comment]
And I noticed that the bitmap pixel data is not always 15552 bytes long...
If I log the bitmap of the `Page Down` button, I get the following:
```
<Buffer 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f ... 15502 more bytes>
<Buffer 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f ... 15502 more bytes>
<Buffer 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f ... 11467 more bytes>
<Buffer 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f ... 15502 more bytes>
<Buffer 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f ... 11467 more bytes>
<Buffer 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f 0f ... 11440 more bytes>
```
Somehow, a few bytes are missing sometimes. I don't know, why. That happens when I click the `Page Down` and `Page Up` buttons and as far as I can tell, it only happens for those buttons (including `Page Num`), all the others work fine.

**Comment by Julusian** [comment]
I don't think that changing the render size will be a simple thing to do. All of the rendering code is written heavily around being 72px, as well as modules which do their own pixel buffer generation (which gets passed to companion for displaying on the buttons).
This is something I would like to look at improving, but havent had a chance to.
Some questions I have around this are:
* What range of resolutions should modules be expected to work with?
* How can we/should we handle devices wanting different resolutions? Should we render once and scale for the rest, or should we render it multiple times?
* Can we reliably get text to be a consistent size across resolutions? We use a 'custom' font with a fixed list of sizes, do we need to switch to proper text rendering to support this

For the buffers being the wrong length, I shall have a look into it but could you check that you arent accidentally trimming the end of the message off? Perhaps the TCP segmenting is causing the last fragment to come later than you expect sometimes? The code is doing nothing special for those images, unless we arent initialising the end of the buffers sufficiently so they get trimmed

**Comment by hrueger** [comment]
The only case I have for different image sizes is that my satellite device is a StreamDeck XL. How do you currently render images for that one? Do you scale them up?

**Comment by Julusian** [comment]
Yes, we currently scale them up. It would be nicer to do them at native res, but this is easier and looks good enough.

Not currently. Ill do some testing and get back to you

**Comment by hrueger** [comment]
@Julusian Did you find any issues with the bitmaps?

And I have one more question (also regarding the exposure of the page number in the API (see above)):

I'd like to trigger actions programatically (either from within a module or externally, that doesn't matter). The user should be able to choose those actions.

If I setup my app to appear as a Satellite Device, I can press buttons programatically. I can also ask the user on what page he put his actions. However, I currently don't know what Page my satellite device is currently on and therefore I can't trigger the buttons...

I would like to avoid to assume that I start on page 1 and just blindly triggering the "PAGE UP" button until my counter is at the desired page because that feels very wrong...

Therefore it would be great if I could just navigate directly to page n via the API, preferably having named pages like you said earlier?

Or even better: Can I trigger actions from other modules programmatically without using a button? Then, I'd probably need to enumerate them, too.

Thanks for your time!

**Comment by Julusian** [comment]
Sorry, I have forgotten to do much looking into this. I have been using this new satellite api with companion-satellite for a few weeks and have not had any corruption or missing updates

If you want to trigger arbitrary buttons, then you could look at using one of the TCP, UDP or OSC apis instead. You dont get the same feedback/styling through those though.
This may or may not be what you are after though. I have been thinking about allowing access to more of the buttons space beyond the selected page through this api, but I havent yet

Yeah, I dont think we will ever expose direct action control over an api. Itll have to expose a lot of stuff, and is likely to break in small ways all the time (modules can change their actions or the options to those actions whenever they like). We may allow for creating buttons which dont live on a page, and can be addressed by a uuid, but not yet.

**Comment by hrueger** [comment]
Hi @Julusian,
I've finally found the bug with the partial buffers. It was actually my fault. I was parsing the message using `string.split("=")`. Most of the times, it works fine, but sometimes the base64-encoded buffer ends with two `=` and that caused problems. So never mind!

I guess, we can close this then?

**Comment by depili** [comment]
I'm thinking of connecting a skaarhoj controller using the satellite api. For future using such controllers it would be really nice to be able to define surfaces with more buttons and more buttons per row. In this case I have a master key 36, with three rows of 12 buttons. Will probably need to do a hacky solution where I register two satellite devices with one of them being on a different page to get all 36 buttons assigned

**Comment by Julusian** [comment]
@depili For now I would suggest splitting it to appear as two devices. Make sure to think about button layout, so that it is easy for the person configuring to understand how they are mapped such as a 8x3 and a 4x3 or two 6x3, will be much easier for the user to map actions to than reading all 32 buttons from one page and 4 from the next.

The xkeys devices have a similar issue in companion currently, and are spanning pages in a hacky fashion, and I am thinking I will need to rework it to do just this so that they can be supported in companion-satellite.

I do intend to support larger pages, but I dont have a timeframe for when it will happen so probably best to not wait for it

**Comment by Julusian** [comment]
The remaining parts of this are also asked for by #2905
